test/lib/ufe/testDeleteCmd.py

```diff
@@ -310,10 +321,240 @@ def testDeleteArgs(self):
+    def testDeleteRestrictionSameLayerDef(self):
+        '''Restrict deleting USD node. Cannot delete a prim defined on another layer.'''
+        
+        # Create our UFE notification observer
+        ufeObs = TestObserver()
+        ufe.Scene.addObserver(ufeObs)
+
+        # select a USD object.
+        mayaPathSegment = mayaUtils.createUfePathSegment('|transform1|proxyShape1')
+        usdPathSegment = usdUtils.createUfePathSegment('/Room_set/Props/Ball_35')
+        ball35Path = ufe.Path([mayaPathSegment, usdPathSegment])
+        ball35PathStr = ','.join(
+            [str(segment) for segment in ball35Path.segments])
+        
+        ball35Item = ufe.Hierarchy.createItem(ball35Path)
+        
+        ufe.GlobalSelection.get().append(ball35Item)
+        
+        # get the USD stage
+        stage = mayaUsd.ufe.getStage(str(mayaPathSegment))
+        
+        # check GetLayerStack behavior
+        self.assertEqual(stage.GetLayerStack()[0], stage.GetSessionLayer())
+        self.assertEqual(stage.GetEditTarget().GetLayer(), stage.GetRootLayer())
+
+        cmds.delete()    
+        
+        self.assertEqual(ufeObs.nbDeleteNotif(), 0)
+
+    @unittest.skipUnless(ufeUtils.ufeFeatureSetVersion() >= 2, 'testDeleteProxyShape only available in UFE v2 or greater.') 
+    def testDeleteRestrictionHasSpecs(self):
+        '''Restrict deleting USD node. Cannot delete a node that doesn't contribute to the final composed prim'''
+
+        cmds.file(new=True, force=True)
 
+        # open appleBite.ma scene in testSamples
+        mayaUtils.openAppleBiteScene()
+
+        # Create our UFE notification observer
+        ufeObs = TestObserver()
+        ufe.Scene.addObserver(ufeObs)
+
+        # clear selection to start off
+        cmds.select(clear=True)
+        
+        # select a USD object.
+        mayaPathSegment = mayaUtils.createUfePathSegment('|Asset_flattened_instancing_and_class_removed_usd|Asset_flattened_instancing_and_class_removed_usdShape')
+        usdPathSegment = usdUtils.createUfePathSegment('/apple/payload/geo')
+        geoPath = ufe.Path([mayaPathSegment, usdPathSegment])
+        geoItem = ufe.Hierarchy.createItem(geoPath)
+        globalSn = ufe.GlobalSelection.get()        
+        globalSn.append(geoItem)
+        self.assertEqual(len(ufe.GlobalSelection.get()), 1)
+
+        cmds.delete()
+        # not deleted
+        self.assertEqual(ufeObs.nbDeleteNotif(), 0)
+
+        # Clear the selection.
+
+        globalSn.clear()
+        self.assertTrue(globalSn.empty())
+
+        usdPathSegment = usdUtils.createUfePathSegment('/apple/payload')
+        geoPath = ufe.Path([mayaPathSegment, usdPathSegment])
+        geoItem = ufe.Hierarchy.createItem(geoPath)
+
+        globalSn.append(geoItem)
+        self.assertEqual(len(ufe.GlobalSelection.get()), 1)
+
+        cmds.delete()
+        #deleted
+        self.assertEqual(ufeObs.nbDeleteNotif(), 1)
```

Comment:
I'm not sure if you already have a test for it...but would be good to test undo and redo for USD objects deletion.

Reply:
Here I'd go beyond checking the number of notifications.  I would ask the parent if it still has this child.